node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java

```diff
@@ -0,0 +1,161 @@
+
+    private static final Logger logger = Logger.getLogger(InfraDeployerImpl.class.getName());
+
+    private final NodeRepository nodeRepository;
+    private final Provisioner provisioner;
+    private final InfrastructureVersions infrastructureVersions;
+    private final DuperModelInfraApi duperModel;
+    private final Map<ApplicationId, InfraApplicationApi> infraApplicationApiByApplicationId;
+
+    @Inject
+    public InfraDeployerImpl(NodeRepository nodeRepository, Provisioner provisioner, DuperModelInfraApi duperModel) {
+        this.nodeRepository = nodeRepository;
+        this.provisioner = provisioner;
+        this.infrastructureVersions = nodeRepository.infrastructureVersions();
+        this.duperModel = duperModel;
+        this.infraApplicationApiByApplicationId = duperModel.getSupportedInfraApplications().stream()
+                .collect(Collectors.toUnmodifiableMap(
+                        InfraApplicationApi::getApplicationId,
+                        Function.identity()));
+    }
+
+
+    @Override
+    public Optional<Deployment> getDeployment(ApplicationId application) {
+        return Optional.ofNullable(infraApplicationApiByApplicationId.get(application)).map(InfraDeployment::new);
+    }
+
+    @Override
+    public List<ApplicationId> getSupportedInfraApplications() {
+        return List.copyOf(infraApplicationApiByApplicationId.keySet());
+    }
+
+    private class InfraDeployment implements Deployment {
+        private final InfraApplicationApi application;
+
+        private boolean prepared = false;
+        private List<Node> candidateNodes;
+        private List<HostSpec> hostSpecs;
+
+        private InfraDeployment(InfraApplicationApi application) {
+            this.application = application;
+        }
+
+        @Override
+        public void prepare() {
+            if (prepared) return;
+
+            try (Mutex lock = nodeRepository.lock(application.getApplicationId())) {
+                NodeType nodeType = application.getCapacity().type();
+
+                Optional<Version> targetVersion = infrastructureVersions.getTargetVersionFor(nodeType);
+                if (targetVersion.isEmpty()) {
+                    logger.log(LogLevel.DEBUG, "No target version set for " + nodeType + ", removing application");
+                    removeApplication(application.getApplicationId());
+                    return;
+                }
+
+                candidateNodes = nodeRepository
+                        .getNodes(nodeType, Node.State.ready, Node.State.reserved, Node.State.active, Node.State.inactive);
+                if (candidateNodes.isEmpty()) {
+                    logger.log(LogLevel.DEBUG, "No nodes to provision for " + nodeType + ", removing application");
+                    removeApplication(application.getApplicationId());
+                    return;
+                }
+
+                if (!candidateNodes.stream().allMatch(node ->
```

Comment:
Consider extracting a method for this condition. Maybe something like `if (!activeNodesOn(version, candidates))`?